buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+        "\\W+public static final Version V_(\\d+)_(\\d+)_(\\d+)(_alpha\\d+|_beta\\d+|_rc\\d+)? .*"
+    );
+
+    private final Version currentVersion;
+    private final Map<Integer, List<Version>> groupByMajor;
+
+    public class UnreleasedVersionDescription {
+        private final Version version;
+        private final String branch;
+        private final String gradleProjectName;
+
+        UnreleasedVersionDescription(Version version, String branch, String gradleProjectName) {
+            this.version = version;
+            this.branch = branch;
+            this.gradleProjectName = gradleProjectName;
+        }
+
+        public Version getVersion() {
+            return version;
+        }
+
+        public String getBranch() {
+            return branch;
+        }
+
+        public String getGradleProjectName() {
+            return gradleProjectName;
+        }
+    }
+
+    public VersionCollection(List<String> versionLines) {
+        this(versionLines, VersionProperties.getElasticsearch());
+    }
+
+    protected VersionCollection(List<String> versionLines, Version currentVersionProperty) {
+        groupByMajor = versionLines.stream()
+            .map(LINE_PATTERN::matcher)
+            .filter(Matcher::matches)
+            .map(match -> new Version(
+                Integer.parseInt(match.group(1)),
+                Integer.parseInt(match.group(2)),
+                Integer.parseInt(match.group(3)),
+                (match.group(4) == null ? "" : match.group(4)).replace('_', '-'),
+                false
+            ))
+            .filter(version -> version.getSuffix().isEmpty() || version.equals(currentVersionProperty))
+            .collect(Collectors.groupingBy(Version::getMajor, Collectors.toList()));
+
+        if (groupByMajor.isEmpty()) {
+            throw new IllegalArgumentException("Could not parse any versions");
+        }
+
+        currentVersion = getLatestVersionByKey(
+            groupByMajor,
+            groupByMajor.keySet().stream().max(Integer::compareTo)
+                .orElseThrow(() -> new IllegalStateException("Unexpected number of versions in collection"))
+        );
+
+        assertCurrentVersionMatchesParsed(currentVersionProperty);
+
+        assertNoOlderThanTwoMajors();
+
+        markUnreleasedAsSnapshot();
+    }
+
+    private void markUnreleasedAsSnapshot() {
+        getUnreleased()
+            .forEach(unreleased -> {
+                groupByMajor.get(unreleased.getMajor()).remove(unreleased);
+                groupByMajor.get(unreleased.getMajor()).add(
+                    new Version(
+                        unreleased.getMajor(), unreleased.getMinor(), unreleased.getRevision(),
+                        unreleased.getSuffix(), true
+                    )
+                );
+            });
+    }
+
+    private void assertNoOlderThanTwoMajors() {
+        Set<Integer> majors = groupByMajor.keySet();
+        if (majors.size() != 2) {
```

Comment:
This will break when we bump master to the next major release, since old version constants will not yet be removed. I think you can detect this, though, and allow only when current is master.

Reply:
I'm assuming it's a lot of work to remove those that's why it's not  done at the same time right ?